The beforeinput event should fire before textInput

sideshowbarker opened the pull request:
#### 7da094e1da191c5d59b9abe0125e4bd9746b9ec2
<pre>
The beforeinput event should fire before textInput
<a href="https://bugs.webkit.org/show_bug.cgi?id=268988">https://bugs.webkit.org/show_bug.cgi?id=268988</a>

Reviewed by Wenson Hsieh and Ryosuke Niwa.

This change makes WebKit fire the beforeinput &amp; textInput events in the
order (beforeinput first, textInput after) conforming to UI Events spec
requirements at <a href="https://github.com/w3c/uievents/pull/362">https://github.com/w3c/uievents/pull/362</a> and in the spec at
<a href="https://w3c.github.io/uievents/event-algo.html#fire%20key%20input%20events">https://w3c.github.io/uievents/event-algo.html#fire%20key%20input%20events</a>
and consistent with the order in which the events are fired in Blink.

Otherwise, without this change, WebKit fires the events in an order
(textInput first, beforeinput after) that breaks conformance with
the spec requirements, and that breaks compatibility with Blink.

Note also that this change makes WebKit conform to the requirements in
<a href="https://w3c.github.io/uievents/event-algo.html#fire%20key%20input%20events">https://w3c.github.io/uievents/event-algo.html#fire%20key%20input%20events</a>,
<a href="https://w3c.github.io/uievents/event-algo.html#handle%20native%20paste">https://w3c.github.io/uievents/event-algo.html#handle%20native%20paste</a>, and
<a href="https://w3c.github.io/uievents/event-algo.html#end%20composition">https://w3c.github.io/uievents/event-algo.html#end%20composition</a> — limiting
textInput to being fired only when ending a composition or when the input
type is insertText, insertParagraph, insertLineBreak, or insertFromPaste.

* LayoutTests/editing/execCommand/break-out-of-empty-list-item.html:
* LayoutTests/editing/inserting/typing-space-to-trigger-smart-link.html:
* LayoutTests/editing/pasteboard/paste-text-events-expected.txt:
* LayoutTests/editing/pasteboard/paste-text-events.html:
* LayoutTests/editing/style/highlight-insert-paragraph.html:
* LayoutTests/fast/events/input-events-fired-when-typing-expected.txt:
* LayoutTests/fast/events/input-events-fired-when-typing.html:
* LayoutTests/fast/events/ios/submit-form-target-blank-using-return-key.html:
* LayoutTests/fast/events/onchange-passwordfield.html:
* LayoutTests/fast/events/onchange-searchfield.html:
* LayoutTests/fast/events/onchange-textfield.html:
* LayoutTests/fast/forms/onchange-change-type.html:
* LayoutTests/http/tests/navigation/keyboard-events-during-provisional-navigation-expected.txt:
* LayoutTests/http/tests/navigation/keyboard-events-during-provisional-subframe-navigation-expected.txt:
* LayoutTests/platform/mac-wk2/http/tests/navigation/keyboard-events-during-provisional-navigation-expected.txt:
* LayoutTests/platform/wpe/TestExpectations:
* Source/WebCore/dom/Node.cpp:
(WebCore::Node::defaultEventHandler):
* Source/WebCore/dom/TextEvent.cpp:
(WebCore::TextEvent::TextEvent):
(WebCore::TextEvent::initTextEvent):
(WebCore::TextEvent::createForPlainTextPaste): Deleted.
(WebCore::TextEvent::createForFragmentPaste): Deleted.
(WebCore::TextEvent::createForDictation): Deleted.
* Source/WebCore/dom/TextEvent.h:
* Source/WebCore/editing/AlternativeTextController.cpp:
(WebCore::AlternativeTextController::insertDictatedText):
* Source/WebCore/editing/Editor.cpp:
(WebCore::dispatchTextInputEvent):
(WebCore::Editor::selectionForCommand):
(WebCore::Editor::pasteAsPlainText):
(WebCore::Editor::pasteAsFragment):
(WebCore::dispatchTextInputEvents):
(WebCore::Editor::appliedEditing):
(WebCore::Editor::insertText):
(WebCore::Editor::insertTextForConfirmedComposition):
(WebCore::Editor::insertTextWithoutSendingTextEvent):
(WebCore::Editor::setComposition):
(WebCore::Editor::handleTextEvent): Deleted.
* Source/WebCore/editing/Editor.h:
* Source/WebCore/editing/EditorCommand.cpp:
(WebCore::executeInsertBacktab):
(WebCore::executeInsertLineBreak):
(WebCore::executeInsertNewline):
(WebCore::executeInsertTab):
(WebCore::executeYank):
(WebCore::executeYankAndSelect):
(WebCore::enabledVisibleSelection):
(WebCore::enabledVisibleSelectionAndMark):
(WebCore::enableCaretInEditableText):
(WebCore::enabledInEditableText):
* Source/WebCore/page/EventHandler.cpp:
(WebCore::EventHandler::handleTextInput):
(WebCore::EventHandler::handleTextInputEvent): Deleted.
(WebCore::EventHandler::defaultTextInputEventHandler): Deleted.
* Source/WebCore/page/EventHandler.h:

Canonical link: <a href="https://commits.webkit.org/278971@main">https://commits.webkit.org/278971@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/cbcb3520d6428902fbac6ad92fb21715ccaa8106

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/50312 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/29605 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/2609 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/53571 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/1002 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/52615 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/35863 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/650 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/41042 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/531 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/52411 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/27287 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/43304 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/22146 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/24697 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/557 "Passed tests") | [⏳ 🛠 wpe-cairo ](https://ews-build.webkit.org/#/builders/WPE-Cairo-Build-EWS "Waiting in queue, processing has not started yet") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/46682 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/621 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/55157 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/25408 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/552 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/48447 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/26671 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/43487 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/47479 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/11405 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/27533 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/26401 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rniwa commented on Source/WebCore/editing/Editor.cpp:
```diff
@@ -1233,8 +1248,15 @@ void Editor::appliedEditing(CompositeEditCommand& command)
+    auto data = command.inputEventData();
 
-    notifyTextFromControls(composition->startingRootEditableElement(), composition->endingRootEditableElement());
+    if (inputType == "insertText"_s || inputType == "insertParagraph"_s || inputType == "insertLineBreak"_s)
```
Why are we checking this? That's not mentioned in the commit message?

sideshowbarker replied:
>>     if (inputType == "insertText"_s || inputType == "insertParagraph"_s || inputType == "insertLineBreak"_s)`
> Why are we checking this? That's not mentioned in the commit message?

That’s per step 4.6 of the algorithm at https://w3c.github.io/uievents/event-algo.html#fire%20key%20input%20events:
> If _inputType_ is not one of "insertText", "insertParagraph" or "insertLineBreak", then set _shouldFireTextInput_ to false.

I’ve updated the commit message to explicitly mention that.

rniwa commented on Source/WebCore/page/EventHandler.cpp:
```diff
@@ -4416,17 +4416,31 @@ bool EventHandler::handleTextInputEvent(const String& text, Event* underlyingEve
     Ref frame = m_frame.get();
 
+    auto isEnterKeypress = [underlyingEvent]()
+    {
```
Nit: { should appear on the same line as `[underlyingEvent]()`.

rniwa commented on Source/WebCore/page/EventHandler.cpp:
```diff
@@ -4416,17 +4416,31 @@ bool EventHandler::handleTextInputEvent(const String& text, Event* underlyingEve
         return false;
 
+    // The UI Events spec requires that textInput not be fired until after beforeinput is fired. So we create
+    // a textInput event here, but stop its propagation before dispatching it. And then we dispatch textInput
```
Why are we firing textInput without propagation here? What's the point of this code?

sideshowbarker replied:
> ```
>    event->stopPropagation();
>    target->dispatchEvent(event);
> ```
> Why are we firing textInput without propagation here?

Fixed now — by instead directly calling `Editor::insertTextWithoutSendingTextEvent` with a nullptr for the `triggeringEvent` arg (rather than trying to preserve the existing `target->dispatchEvent(event)` call).

> What's the point of this code?

Per the _“If the event handler changed the selection, we may want to use a different selection that is contained in the event target”_ comment in the `Editor::insertTextWithoutSendingTextEvent`code (which we end up in after `EventDispatcher::dispatchEvent`), I’d been thinking the logic there for getting the selection still needed to check the target — and so, we needed to pass down an actual triggering event that it could get the target from. But if the event handler here isn’t changing the selection, then that code doesn’t need to check the target, and so doesn’t need an actual event to get the target from.

sideshowbarker replied:
> Fixed now — by instead directly calling `Editor::insertTextWithoutSendingTextEvent` with a nullptr for the `triggeringEvent` arg (rather than trying to preserve the existing `target->dispatchEvent(event)` call).

Seems I spoke too soon. That replacement  causes [regressions with several test cases](https://ews-build.webkit.org/#/builders/36/builds/41661) – including the `fast/forms/focus-change-on-keypress.html` test — and I recall now seeing those same regressions a few weeks earlier when I was first trying out solutions for avoiding the event dispatch.

> > What's the point of this code?
> 
> …in the `Editor::insertTextWithoutSendingTextEvent`code… I’d been thinking the logic there for getting the selection still needed to check the target… but if the event handler here isn’t changing the selection, then that code doesn’t need to check the target

Seems I may have been wrong about that not being needed. I haven’t yet looked in detail at the exact causes of all the regressions, but so far I did step through the `fast/forms/focus-change-on-keypress.html` execution — and the cause in that case does in fact seem to be: we’re skipping the check of the target to get the selection.

So for now, I have reverted to what I had earlier: preserving the `target->dispatchEvent(event)` call — so that the selection-setting code in `Editor::insertTextWithoutSendingTextEvent` can continue working as expected — but then stopping propagation of the event.

I realize that’s kludgy but it’s not clear to me how to otherwise avoid the regressions and get the behavior conforming without reworking a bunch more of the existing code.

But maybe I’m overlooking something and there’s some simpler change we could make to avoid the kludge.

aprotyas commented on Source/WebCore/page/EventHandler.cpp:
```diff
@@ -4416,17 +4416,31 @@ bool EventHandler::handleTextInputEvent(const String& text, Event* underlyingEve
 
     Ref frame = m_frame.get();
 
+    auto isEnterKeypress = [underlyingEvent]()
```
Nit: You can omit an empty parameter list.
```suggestion
    auto isEnterKeypress = [underlyingEvent]
```

rniwa reviewed (approved):
Hopefully no compat issues.

sideshowbarker committed:
The beforeinput event should fire before textInput
https://bugs.webkit.org/show_bug.cgi?id=268988

Reviewed by Wenson Hsieh and Ryosuke Niwa.

This change makes WebKit fire the beforeinput & textInput events in the
order (beforeinput first, textInput after) conforming to UI Events spec
requirements at https://github.com/w3c/uievents/pull/362 and in the spec at
https://w3c.github.io/uievents/event-algo.html#fire%20key%20input%20events
and consistent with the order in which the events are fired in Blink.

Otherwise, without this change, WebKit fires the events in an order
(textInput first, beforeinput after) that breaks conformance with
the spec requirements, and that breaks compatibility with Blink.

Note also that this change makes WebKit conform to the requirements in
https://w3c.github.io/uievents/event-algo.html#fire%20key%20input%20events,
https://w3c.github.io/uievents/event-algo.html#handle%20native%20paste, and
https://w3c.github.io/uievents/event-algo.html#end%20composition — limiting
textInput to being fired only when ending a composition or when the input
type is insertText, insertParagraph, insertLineBreak, or insertFromPaste.

* LayoutTests/editing/execCommand/break-out-of-empty-list-item.html:
* LayoutTests/editing/inserting/typing-space-to-trigger-smart-link.html:
* LayoutTests/editing/pasteboard/paste-text-events-expected.txt:
* LayoutTests/editing/pasteboard/paste-text-events.html:
* LayoutTests/editing/style/highlight-insert-paragraph.html:
* LayoutTests/fast/events/input-events-fired-when-typing-expected.txt:
* LayoutTests/fast/events/input-events-fired-when-typing.html:
* LayoutTests/fast/events/ios/submit-form-target-blank-using-return-key.html:
* LayoutTests/fast/events/onchange-passwordfield.html:
* LayoutTests/fast/events/onchange-searchfield.html:
* LayoutTests/fast/events/onchange-textfield.html:
* LayoutTests/fast/forms/onchange-change-type.html:
* LayoutTests/http/tests/navigation/keyboard-events-during-provisional-navigation-expected.txt:
* LayoutTests/http/tests/navigation/keyboard-events-during-provisional-subframe-navigation-expected.txt:
* LayoutTests/platform/mac-wk2/http/tests/navigation/keyboard-events-during-provisional-navigation-expected.txt:
* LayoutTests/platform/wpe/TestExpectations:
* Source/WebCore/dom/Node.cpp:
(WebCore::Node::defaultEventHandler):
* Source/WebCore/dom/TextEvent.cpp:
(WebCore::TextEvent::TextEvent):
(WebCore::TextEvent::initTextEvent):
(WebCore::TextEvent::createForPlainTextPaste): Deleted.
(WebCore::TextEvent::createForFragmentPaste): Deleted.
(WebCore::TextEvent::createForDictation): Deleted.
* Source/WebCore/dom/TextEvent.h:
* Source/WebCore/editing/AlternativeTextController.cpp:
(WebCore::AlternativeTextController::insertDictatedText):
* Source/WebCore/editing/Editor.cpp:
(WebCore::dispatchTextInputEvent):
(WebCore::Editor::selectionForCommand):
(WebCore::Editor::pasteAsPlainText):
(WebCore::Editor::pasteAsFragment):
(WebCore::dispatchTextInputEvents):
(WebCore::Editor::appliedEditing):
(WebCore::Editor::insertText):
(WebCore::Editor::insertTextForConfirmedComposition):
(WebCore::Editor::insertTextWithoutSendingTextEvent):
(WebCore::Editor::setComposition):
(WebCore::Editor::handleTextEvent): Deleted.
* Source/WebCore/editing/Editor.h:
* Source/WebCore/editing/EditorCommand.cpp:
(WebCore::executeInsertBacktab):
(WebCore::executeInsertLineBreak):
(WebCore::executeInsertNewline):
(WebCore::executeInsertTab):
(WebCore::executeYank):
(WebCore::executeYankAndSelect):
(WebCore::enabledVisibleSelection):
(WebCore::enabledVisibleSelectionAndMark):
(WebCore::enableCaretInEditableText):
(WebCore::enabledInEditableText):
* Source/WebCore/page/EventHandler.cpp:
(WebCore::EventHandler::handleTextInput):
(WebCore::EventHandler::handleTextInputEvent): Deleted.
(WebCore::EventHandler::defaultTextInputEventHandler): Deleted.
* Source/WebCore/page/EventHandler.h:

Canonical link: https://commits.webkit.org/278971@main
